File: 2020/lib/day_06.ex
defmodule Day06 do
  # Part 1 and part 2 could probably both be consolidated to use MapSet.
  def part1(input) do
    input
    |> parse()
    |> Enum.map(&String.replace(&1, "\n", ""))
    |> Enum.map(&String.graphemes/1)
    |> Enum.map(&Enum.uniq/1)
    |> List.flatten()
    |> Enum.count()
  end

  def part2(input) do
    input
    |> parse()
    |> Enum.map(fn groups ->
      groups
      |> String.split("\n")
      |> Enum.map(&(String.graphemes(&1) |> MapSet.new()))
      |> Enum.reduce(&MapSet.intersection(&1, &2))
      |> Enum.count()
    end)
    |> Enum.sum()
  end

  defp parse(input) do
    input
    |> Helpers.parse_to_list("\n\n")
  end
end

File: 2021/lib/day_06.rb
class Day06
  def part1(input)
    spawn(initial_timers(input), 80)
  end

  def part2(input)
    spawn(initial_timers(input), 256)
  end

  def spawn(timers, days)
    days
      .times
      .each_with_object(timers) do |_, t|
        t.rotate!
        t[6] += t[8]
      end
      .sum
  end

  def initial_timers(input)
    input
      .split(",")
      .map(&:to_i)
      .each_with_object(Array.new(9, 0)) { |i, t| t[i] += 1 }
  end
end
